File: src/tools/browser.py
"""Browser management tools.

Ported from src/mcp-tools/tools/browser.ts
"""

from typing import Any, Literal, Optional

from playwright.async_api import Page
from pydantic import BaseModel, Field

from src.tools.base import ToolResult, async_session_tool

# ============= Tool Argument Schemas =============


class TabsArgs(BaseModel):
    """Arguments for browser_tabs tool."""

    action: Literal["list", "new", "close", "select"] = Field(
        description="Tab operation to perform"
    )
    index: Optional[int] = Field(
        default=None,
        ge=0,
        description="Tab index for close/select operations",
    )
    url: Optional[str] = Field(
        default=None,
        description="URL for new tab",
    )


class ResizeArgs(BaseModel):
    """Arguments for browser_resize tool."""

    width: int = Field(
        ge=100,
        le=10000,
        description="Window width in pixels",
    )
    height: int = Field(
        ge=100,
        le=10000,
        description="Window height in pixels",
    )


class HandleDialogArgs(BaseModel):
    """Arguments for browser_handle_dialog tool."""

    accept: bool = Field(
        default=True,
        description="Whether to accept (OK) or dismiss (Cancel) the dialog",
    )
    prompt_text: Optional[str] = Field(
        default=None,
        description="Text to enter for prompt dialogs",
    )


class ReloadArgs(BaseModel):
    """Arguments for browser_reload tool."""

    force: bool = Field(
        default=False,
        description="Whether to force reload from server (bypass cache)",
    )
    wait_until: Literal["load", "domcontentloaded", "networkidle"] = Field(
        default="load",
        description="When to consider reload complete",
    )


class CloseArgs(BaseModel):
    """Arguments for browser_close tool."""

    close_browser: bool = Field(
        default=False,
        description="Whether to close entire browser (true) or just current page (false)",
    )


# ============= Tool Functions =============


@async_session_tool
async def browser_tabs(
    action: str,
    index: Optional[int] = None,
    url: Optional[str] = None,
    page: Page = None,
) -> str:
    """List, create, close, or select browser tabs."""
    try:
        browser_context = page.context

        if action == "list":
            pages = browser_context.pages
            current_index = pages.index(page)
            tabs = []
            for i, p in enumerate(pages):
                tabs.append(
                    {
                        "index": i,
                        "url": p.url,
                        "title": await p.title(),
                        "active": i == current_index,
                    }
                )
            return ToolResult(
                success=True,
                content=f"Browser tabs ({len(tabs)} total):\n{tabs}",
                data={"tabs": tabs},
            ).to_string()

        if action == "new":
            new_page = await browser_context.new_page()
            if url:
                await new_page.goto(url)
            return ToolResult(
                success=True,
                content=f"Created new tab{f' and navigated to {url}' if url else ''}",
                data={"url": url},
            ).to_string()

        if action == "close":
            if index is not None:
                pages = browser_context.pages
                if 0 <= index < len(pages):
                    await pages[index].close()
                    return ToolResult(
                        success=True,
                        content=f"Closed tab at index {index}",
                        data={"index": index},
                    ).to_string()
                return ToolResult(
                    success=False,
                    content=f"Invalid tab index: {index}",
                ).to_string()
            await page.close()
            return ToolResult(success=True, content="Closed current tab").to_string()

        if action == "select":
            if index is None:
                return ToolResult(
                    success=False,
                    content="Tab index required for select action",
                ).to_string()
            pages = browser_context.pages
            if 0 <= index < len(pages):
                await pages[index].bring_to_front()
                return ToolResult(
                    success=True,
                    content=f"Switched to tab at index {index}",
                    data={"index": index},
                ).to_string()
            return ToolResult(
                success=False,
                content=f"Invalid tab index: {index}",
            ).to_string()

        return ToolResult(success=False, content=f"Unknown action: {action}").to_string()
    except Exception as exc:
        return ToolResult(
            success=False,
            content=f"Tab action failed: {exc}",
        ).to_string()


@async_session_tool
async def browser_resize(width: int, height: int, page: Page = None) -> str:
    """Resize the browser window to specified dimensions."""
    try:
        await page.set_viewport_size({"width": width, "height": height})
        return ToolResult(
            success=True,
            content=f"Browser window resized to {width}x{height}",
            data={"width": width, "height": height},
        ).to_string()
    except Exception as exc:
        return ToolResult(
            success=False,
            content=f"Failed to resize browser window: {exc}",
        ).to_string()


@async_session_tool
async def browser_handle_dialog(
    accept: bool = True,
    prompt_text: Optional[str] = None,
    page: Page = None,
) -> str:
    """Handle JavaScript dialogs (alert, confirm, prompt)."""
    try:
        def handler(dialog: Any) -> None:
            if prompt_text and dialog.type == "prompt":
                dialog.accept(prompt_text)
            elif accept:
                dialog.accept()
            else:
                dialog.dismiss()

        page.once("dialog", handler)

        return ToolResult(
            success=True,
            content=(
                f"Dialog handler configured (accept: {accept}"
                f"{f', prompt: {prompt_text}' if prompt_text else ''})"
            ),
            data={"accept": accept, "prompt_text": prompt_text},
        ).to_string()
    except Exception as exc:
        return ToolResult(
            success=False,
            content=f"Failed to configure dialog handler: {exc}",
        ).to_string()


@async_session_tool
async def browser_reload(
    force: bool = False,
    wait_until: str = "load",
    page: Page = None,
) -> str:
    """Reload the current page."""
    try:
        await page.reload(wait_until=wait_until, timeout=30000)
        page_title = await page.title()
        return ToolResult(
            success=True,
            content=f"Page reloaded{' (forced)' if force else ''}. Title: {page_title}",
            data={"title": page_title, "force": force},
        ).to_string()
    except Exception as exc:
        return ToolResult(
            success=False,
            content=f"Failed to reload page: {exc}",
        ).to_string()


@async_session_tool
async def browser_close(close_browser: bool = False, page: Page = None) -> str:
    """Close the browser or current page."""
    try:
        if close_browser:
            browser = page.context.browser
            if browser:
                await browser.close()
                return ToolResult(success=True, content="Browser closed").to_string()
            await page.context.close()
            return ToolResult(success=True, content="Browser context closed").to_string()
        await page.close()
        return ToolResult(success=True, content="Current page closed").to_string()
    except Exception as exc:
        return ToolResult(
            success=False,
            content=f"Failed to close browser/page: {exc}",
        ).to_string()

File: src/mcp-tools/tools/browser.ts
import { BaseMCPTool } from '../core/base-tool';
import { MCPContext, MCPToolResult } from '../core/types';

export class TabsTool extends BaseMCPTool {
  definition = {
    name: 'browser_tabs',
    description: 'List, create, close, or select browser tabs',
    inputSchema: {
      type: 'object' as const,
      properties: {
        action: { type: 'string', description: 'Tab operation to perform', enum: ['list', 'new', 'close', 'select'] },
        index: { type: 'number', description: 'Tab index for close/select operations', minimum: 0 },
        url: { type: 'string', description: 'URL for new tab', format: 'uri' }
      },
      required: ['action']
    },
    metadata: { tags: ['browser', 'tabs', 'management'] }
  };

  protected async _execute(args: any, context: MCPContext): Promise<MCPToolResult> {
    const { action, index, url } = args;
    const { page, context: browserContext } = context;

    switch (action) {
      case 'list': {
        const pages = browserContext.pages();
        const currentIndex = pages.indexOf(page);
        const tabs = await Promise.all(pages.map(async (p: any, i: number) => ({
          index: i,
          url: p.url(),
          title: await p.title(),
          active: i === currentIndex
        })));

        return {
          content: [{ type: 'text', text: `Browser tabs (${tabs.length} total):\n${JSON.stringify(tabs, null, 2)}` }]
        };
      }

      case 'new': {
        const newPage = await browserContext.newPage();
        if (url) {
          await newPage.goto(url);
        }
        return this.textResult(`Created new tab${url ? ` and navigated to ${url}` : ''}`);
      }

      case 'close': {
        if (index !== undefined) {
          const pages = browserContext.pages();
          if (index >= 0 && index < pages.length) {
            await pages[index].close();
            return this.textResult(`Closed tab at index ${index}`);
          }
          return this.errorResult(`Invalid tab index: ${index}`);
        }
        await page.close();
        return this.textResult('Closed current tab');
      }

      case 'select': {
        if (index !== undefined) {
          const pages = browserContext.pages();
          if (index >= 0 && index < pages.length) {
            await pages[index].bringToFront();
            return this.textResult(`Switched to tab at index ${index}`);
          }
          return this.errorResult(`Invalid tab index: ${index}`);
        }
        return this.errorResult('Tab index required for select action');
      }

      default:
        return this.errorResult(`Unknown action: ${action}`);
    }
  }
}

export class ResizeTool extends BaseMCPTool {
  definition = {
    name: 'browser_resize',
    description: 'Resize the browser window to specified dimensions',
    inputSchema: {
      type: 'object' as const,
      properties: {
        width: { type: 'number', description: 'Window width in pixels', minimum: 100, maximum: 10000 },
        height: { type: 'number', description: 'Window height in pixels', minimum: 100, maximum: 10000 }
      },
      required: ['width', 'height']
    },
    metadata: { tags: ['browser', 'window', 'resize'] }
  };

  protected async _execute(args: any, context: MCPContext): Promise<MCPToolResult> {
    const { width, height } = args;
    const { page } = context;
    await page.setViewportSize({ width, height });
    return this.textResult(`Browser window resized to ${width}x${height}`);
  }
}

export class HandleDialogTool extends BaseMCPTool {
  definition = {
    name: 'browser_handle_dialog',
    description: 'Handle JavaScript dialogs (alert, confirm, prompt)',
    inputSchema: {
      type: 'object' as const,
      properties: {
        accept: { type: 'boolean', description: 'Whether to accept (OK) or dismiss (Cancel) the dialog', default: true },
        promptText: { type: 'string', description: 'Text to enter for prompt dialogs' }
      },
      required: []
    },
    metadata: { tags: ['browser', 'dialog', 'alert'] }
  };

  protected async _execute(args: any, context: MCPContext): Promise<MCPToolResult> {
    const { accept = true, promptText } = args;
    const { page } = context;

    page.once('dialog', async (dialog: any) => {
      if (promptText && dialog.type() === 'prompt') {
        await dialog.accept(promptText);
      } else if (accept) {
        await dialog.accept();
      } else {
        await dialog.dismiss();
      }
    });

    return this.textResult(`Dialog handler configured (accept: ${accept}${promptText ? `, prompt: ${promptText}` : ''})`);
  }
}

export class ReloadTool extends BaseMCPTool {
  definition = {
    name: 'browser_reload',
    description: 'Reload the current page',
    inputSchema: {
      type: 'object' as const,
      properties: {
        force: { type: 'boolean', description: 'Whether to force reload from server (bypass cache)', default: false },
        waitUntil: { type: 'string', description: 'When to consider reload complete', enum: ['load', 'domcontentloaded', 'networkidle'], default: 'load' }
      },
      required: []
    },
    metadata: { tags: ['browser', 'reload', 'refresh'] }
  };

  protected async _execute(args: any, context: MCPContext): Promise<MCPToolResult> {
    const { force = false, waitUntil = 'load' } = args;
    const { page } = context;

    await page.reload({ waitUntil: waitUntil as any, timeout: 30000 });
    const pageTitle = await page.title();

    return this.textResult(`Page reloaded${force ? ' (forced)' : ''}. Title: ${pageTitle}`);
  }
}

export class CloseTool extends BaseMCPTool {
  definition = {
    name: 'browser_close',
    description: 'Close the browser or current page',
    inputSchema: {
      type: 'object' as const,
      properties: {
        closeBrowser: { type: 'boolean', description: 'Whether to close entire browser (true) or just current page (false)', default: false }
      },
      required: []
    },
    metadata: { tags: ['browser', 'close'], destructive: true }
  };

  protected async _execute(args: any, context: MCPContext): Promise<MCPToolResult> {
    const { closeBrowser = false } = args;
    const { page, context: browserContext } = context;

    if (closeBrowser) {
      await browserContext.browser()?.close();
      return this.textResult('Browser closed');
    } else {
      await page.close();
      return this.textResult('Current page closed');
    }
  }
}